README.md
# django-test-mixins

Additional assertions and test conveniences for testing django
sites.

[![Build Status](https://travis-ci.org/Wilfred/django-test-mixins.svg)](https://travis-ci.org/Wilfred/django-test-mixins)
[![Coverage Status](https://img.shields.io/coveralls/Wilfred/django-test-mixins.svg)](https://coveralls.io/r/Wilfred/django-test-mixins)
[![Latest Version](https://img.shields.io/pypi/v/django_test_mixins.svg)](https://pypi.python.org/pypi/django_test_mixins/)
[![License](https://img.shields.io/pypi/l/django_test_mixins.svg)](https://pypi.python.org/pypi/django_test_mixins/)

<!-- markdown-toc start - Don't edit this section. Run M-x markdown-toc-generate-toc again -->
**Table of Contents**

- [django-test-mixins](#django-test-mixins)
    - [Usage](#usage)
        - [HttpCodeTestCase](#httpcodetestcase)
        - [EmptyCacheTestCase](#emptycachetestcase)
        - [FormValidationTestCase](#formvalidationtestcase)
        - [RedirectTestCase](#redirecttestcase)
        - [Combining test cases](#combining-test-cases)
    - [Uploading to PyPI](#uploading-to-pypi)

<!-- markdown-toc end -->

## Usage

All classes inherit from
[django.test.TestCase](https://docs.djangoproject.com/en/dev/topics/testing/overview/#django.test.TestCase)
so you can access the Django assertions and the HTTP client as usual.

django-test-mixins is
[available on PyPI](https://pypi.python.org/pypi/django_test_mixins). Installation
is simply a matter of:

    $ pip install django_test_mixins

### HttpCodeTestCase

`HttpCodeTestCase` provides the following assertions for exact HTTP responses:

* `assertHttpOK(response)` (200)
* `assertHttpCreated(response)` (201)
* `assertHttpBadRequest(response)` (400)
* `assertHttpUnauthorized(response)` (401)
* `assertHttpForbidden(response)` (403)
* `assertHttpNotFound(response)` (404)
* `assertHttpMethodNotAllowed(response)` (405)

It also provides the following assertions for groups of HTTP
responses:

* `assertHttpRedirect(response, location=None)` (3XX)

Example:

```python
from django.core.urlresolvers import reverse

from django_test_mixins import HttpCodeTestCase


class TestIndex(HttpCodeTestCase):
    def test_home_page(self):
        response = self.client.get(reverse('index'))
        self.assertHttpOK(response)
```

### EmptyCacheTestCase

`EmptyCacheTestCase` provides no extra assertions, but ensures that
every test starts with a empty cache.

### FormValidationTestCase

`FormValidationTestCase` provides the following assertion:

* `assertFormInvalid(response, form_name="form")`

Example:

Assume we have a view that looks like this:

```python
from django.shortcuts import render, redirect

from .forms import HouseForm


def create_house(request):
    if request.method == 'POST':
        form = HouseForm(request.POST)

        if form.is_valid():
            form.save()
            return redirect('index')
    else:
        form = HouseForm()

    return render(request, "create_house.html", {'house_form': form})
```

We can then write a test like this. Note that `form_name` needs to
match the context we rendered the page with (defaults to `"form"`).

```python
from django.core.urlresolvers import reverse

from django_test_mixins import FormValidationTestCase


class TestCreateHouse(FormValidationTestCase):
    def test_create_requires_name(self):
        response = self.client.post(reverse('create_house'), {})
        self.assertFormInvalid(response, form_name="house_form")
```

### RedirectTestCase

`RedirectTestCase` provides the following assertions:

* `assertRedirectsTo(response, expected_url)`

Django has `TestCase.assertRedirects`
([docs](https://docs.djangoproject.com/en/dev/topics/testing/overview/#django.test.SimpleTestCase.assertRedirects))
but this does not support redirects to external URLs because of
limitations in the test client.

`assertRedirectsTo` does not support chained redirects, so you can't
do `self.client.get(SOME_URL, follow=True)`.

Example:

```python
from django.core.urlresolvers import reverse

from django_test_mixins import FormValidationTestCase


class TestIndex(RedirectTestCase):
    def test_home_redirects_to_login(self):
        """If the user isn't logged in, index should redirect to the login page.

        """
        response = self.client.get(reverse('index'))
        self.assertRedirectsTo(response, reverse('log_in'))
```


### Combining test cases

You can freely combine these classes by simply inheriting from
multiple classes.

```python
from django.core.urlresolvers import reverse

from django_test_mixins import FreshCacheTestCase, HttpCodeTestCase


class TestIndex(FreshCacheTestCase, HttpCodeTestCase):
    def test_home_page(self):
        response = self.client.get(reverse('index'))
        self.assertHttpOK(response)
```

## Uploading to PyPI

Releasing a new version is a matter of:

    $ python setup.py sdist upload


CHANGELOG.md
## v1.11

Added Python 3 support.

Tags 1.11.1 and 1.11.2 contain the same code, but were created due to
problems releasing from Travis.

## v1.10

Fixed a bug with `.assertRedirectsTo` where absolute HTTPS URLs were
treated as relative URLs.

## v1.9

Fixed a bug where the `form_name` argument to `assertFormInvalid` was
being ignored.

## v1.8

`HttpCodeTestCase.assertHttpRedirect` can now take a `location`
argument for verifying that redirects go to a specific path.

## v1.7.1

Minor spelling fix in assertion message.

## v1.7

Added `HttpCodeTestCase.assertHttpRedirect`

## v1.6

Added `HttpCodeTestCase.assertHttpUnauthorized`

## v1.5

Added `HttpCodeTestCase.assertHttpMethodNotAllowed`

## v1.4

Added `HttpCodeTestCase.assertHttpCreated`

## v1.3

Added `RedirectTestCase`.

## v1.2

Fixed a crash on `HttpCodeTestCase` assertions when the assertion
failed.

## v1.1

Fixed `HttpCodeTestCase` assertions which were throwing
`AttributeError`.

## v1.0

Initial release.


tests.py
from unittest import main
from mock import Mock, MagicMock

# set up dummy settings
# this needs to happen before we import django_test_mixins
from django.conf import settings
settings.configure(DATABASES={
    'default': {
        'ENGINE': 'django.db.backends.sqlite3',
        'NAME': 'mydatabase'
    }
})

from django.core.cache import cache
from django.test import TestCase
from django_test_mixins import HttpCodeTestCase, FormValidationTestCase, EmptyCacheTestCase


class HttpAssertionsTest(HttpCodeTestCase):
    def test_http_ok(self):
        self.assertHttpOK(Mock(status_code=200))

        with self.assertRaises(AssertionError):
            self.assertHttpOK(Mock(status_code=404))

    def test_http_created(self):
        self.assertHttpCreated(Mock(status_code=201))

        with self.assertRaises(AssertionError):
            self.assertHttpCreated(Mock(status_code=404))

    def test_http_bad_request(self):
        self.assertHttpBadRequest(Mock(status_code=400))

        with self.assertRaises(AssertionError):
            self.assertHttpBadRequest(Mock(status_code=200))

    def test_http_unauthorized(self):
        self.assertHttpUnauthorized(Mock(status_code=401))

        with self.assertRaises(AssertionError):
            self.assertHttpUnauthorized(Mock(status_code=200))

    def test_http_forbidden(self):
        self.assertHttpForbidden(Mock(status_code=403))

        with self.assertRaises(AssertionError):
            self.assertHttpForbidden(Mock(status_code=200))

    def test_http_not_found(self):
        self.assertHttpNotFound(Mock(status_code=404))

        with self.assertRaises(AssertionError):
            self.assertHttpNotFound(Mock(status_code=200))

    def test_http_redirect(self):
        self.assertHttpRedirect(Mock(status_code=301))
        self.assertHttpRedirect(Mock(status_code=302))

        with self.assertRaises(AssertionError):
            self.assertHttpRedirect(Mock(status_code=200))

    def test_http_redirect_location(self):
        mock_response = MagicMock(status_code=302)
        mock_response.__getitem__.return_value = "http://testserver/foo"
        
        self.assertHttpRedirect(mock_response, location="foo")

        with self.assertRaises(AssertionError):
            self.assertHttpRedirect(mock_response, location="bar")


class FormAssertionsTest(FormValidationTestCase):
    def test_form_invalid(self):
        """If we have a request containing a form with .is_valid returning
        False, then the form was indeed invalid.

        """
        mock_form = Mock()
        mock_form.is_valid.return_value = False
        
        mock_request = Mock(context={'form': mock_form}, status_code=200)
        self.assertFormInvalid(mock_request)

    def test_form_invalid_custom_name(self):
        mock_form = Mock()
        mock_form.is_valid.return_value = False
        
        mock_request = Mock(context={'my_form': mock_form}, status_code=200)
        self.assertFormInvalid(mock_request, form_name="my_form")

    def test_form_invalid_fails_on_redirect(self):
        """If the form was valid, the most common scenario is that we were
        redirected and there's no form present.

        """
        mock_request = Mock(status_code=302, context={})

        with self.assertRaises(AssertionError):
            self.assertFormInvalid(mock_request)


class CacheTest(TestCase):
    def test_cache_emptied(self):
        cache.set('foo', 1)
        self.assertEqual(cache.get('foo'), 1, "Cache doesn't seem to be configured.")

        class MyTestCase(EmptyCacheTestCase):
            def runTest(self):
                pass

        test_case = MyTestCase()
        test_case.setUp()

        self.assertEqual(cache.get('foo'), None,
                         "Cache was not emptied during setUp!")


if __name__ == '__main__':
    main()


django_test_mixins.py
from django.test import TestCase
from django.core.cache import cache

try:
    from urlparse import urljoin  # python2
except ImportError:
    from urllib.parse import urljoin  # python3


class HttpCodeTestCase(TestCase):
    # TODO: this should be a private method.
    def assertHttpCode(self, response, code, code_description):
        self.assertEqual(
            response.status_code, code,
            "Expected an HTTP %s (%s) response, but got HTTP %s" %
            (code, code_description, response.status_code))

    def assertHttpOK(self, response):
        self.assertHttpCode(response, 200, "OK")

    def assertHttpCreated(self, response):
        self.assertHttpCode(response, 201, "Created")

    def assertHttpRedirect(self, response, location=None):
        """Assert that we had any redirect status code.

        """
        self.assertTrue(
            300 <= response.status_code < 400,
            "Expected an HTTP 3XX (redirect) response, but got HTTP %s" %
            response.status_code
        )

        if location:
            if location.startswith("http://testserver/"):
                absolute_location = location
            else:
                absolute_location = urljoin("http://testserver/", location)

            self.assertEqual(response['Location'], absolute_location)


    def assertHttpBadRequest(self, response):
        self.assertHttpCode(response, 400, "Bad Request")

    def assertHttpUnauthorized(self, response):
        self.assertHttpCode(response, 401, "Unauthorized")

    def assertHttpForbidden(self, response):
        self.assertHttpCode(response, 403, "Forbidden")

    def assertHttpNotFound(self, response):
        self.assertHttpCode(response, 404, "Not Found")

    def assertHttpMethodNotAllowed(self, response):
        self.assertHttpCode(response, 405, "Method Not Allowed")


class EmptyCacheTestCase(TestCase):
    """Ensure that every test starts with an empty cache."""
    def setUp(self):
        super(EmptyCacheTestCase, self).setUp()
        cache.clear()


class FormValidationTestCase(TestCase):
    def assertFormInvalid(self, response, form_name="form"):
        """Assert that the response contains a form in the context, and that
        the form failed validation. The form is assumed to be in
        context[form_name].

        If the form has validated when it shouldn't, views often
        redirect somewhere, so we also check for HTTP 200.

        """
        form = None
        try:
            if response.context:
                form = response.context[form_name]
        except KeyError:
            pass

        if not form:
            self.fail("Could not find a form in the response.")

        self.assertFalse(form.is_valid(), "Expected form to be invalid, but it was valid.")

        status_code = response.status_code
        self.assertEqual(
            status_code, 200,
            "Expected HTTP 200, but got HTTP %d. "
            "Looks like the form validated when it shouldn't." % status_code)


class RedirectTestCase(TestCase):
    def assertRedirectsTo(self, response, expected_url):
        """Django's assertRedirects doesn't support external URLs, so we roll
        our own here. Note that the test client can't fetch external
        URLs, so we mustn't use fetch=True.

        """
        if response.status_code != 302:
            self.fail("Did not redirect (got HTTP %s instead)." % response.status_code)

        if hasattr(response, "redirect_chain"):
            self.fail("You can't use assertRedirects with follow=True.")

        final_url = response._headers['location'][1]

        if not expected_url.startswith('http://') and not expected_url.startswith('https://'):
            # we were given a relative URL, so convert it
            expected_url = "http://testserver%s" % expected_url

        self.assertEqual(
            final_url, expected_url,
            "Expected to be redirected to %s, but got %s instead." % (expected_url, final_url)
        )
